### PR TITLE
Refactor `PolarBasis2D` ~~> `PolarFT`

### PR DESCRIPTION
**Pull request by j-c-c**
This PR consists of 

- [x] Refactoring `PolarBasis2D`
    - [x] Use only half of the polar Fourier rays everywhere
    - [x] Change to `PolarFT`
    - [x] Replace hardcoded tests with logical based tests
- [x] Refactor `commonline_base` and `commonline_sync` code to work with refactored pft
    - [x] Use rays from [0, `n_theta / 2`) instead of [`n_theta / 2`, `n_theta`)
    - [x] Ensure all methods are following the new pf convention
    - [x] Replace hardcoded tests with parametrized logical based tests  

Note, this PR is to address some issues encountered in a downstream branch for the port of the common-lines method using SDP. It was found that our convention for building the common-lines matrix was causing the SDP method to fail.

### CODE REVIEW

**Comment by garrettwrong** [comment]
Is this guy ready for review? It looks like it was auto triggered, but if your are still working on it I won't put on my todo list quite yet. Thanks

**Comment by j-c-c** [comment]
> Is this guy ready for review? It looks like it was auto triggered, but if your are still working on it I won't put on my todo list quite yet. Thanks

Not quite ready. I mistakenly created this as PR instead of a draft. I reverted to a draft, but it looks like you guys will still get pinged about it.

**Comment by j-c-c** [comment]
Hey @garrettwrong, this is ready for an initial review. I'm still working on fixing up the `estimate_shifts` method. I might be able to squeeze it into this PR, but wanted to get an initial review since I'll be out for a few days.

File: src/aspire/abinitio/commonline_base.py
Comment (garrettwrong) [inline src/aspire/abinitio/commonline_base.py]: why reshape to order F and then Transpose?

File: src/aspire/abinitio/commonline_base.py
Comment (garrettwrong) [inline src/aspire/abinitio/commonline_base.py:267]: Is this to support a unit test or something else? (ie, why is it added?) If it is to support something, maybe add a comment, since it isn't obvious. Thanks!

File: src/aspire/operators/polar_ft.py
Comment (garrettwrong) [inline src/aspire/operators/polar_ft.py]: Probably makes the most sense to implement that symmetry based expansion here as a method(s) of this class.  That way it is easier to test and is less likely to become copy pasta.

File: src/aspire/operators/polar_ft.py
Comment (garrettwrong) [inline src/aspire/operators/polar_ft.py]: I'm only mentioning this because we've spent a lot of time going through this code now.... so viewing it as return on investment....

It is likely much clearer to initially construct `freqs` as 3D (2, ntheta//2, nrad) then reshape it after the construction loop.

You can also just compute the trig components and totally factor out the `* np.arange(nrad)` to a single multiply of `freqs` after reshape if you want to optimize it a little.

File: src/aspire/operators/polar_ft.py
Comment (garrettwrong) [inline src/aspire/operators/polar_ft.py]: just curious, if we raise here, what complains?

File: src/aspire/operators/polar_ft.py
Comment (garrettwrong) [inline src/aspire/operators/polar_ft.py:103]: Similarly like L102, what complains here (just curious, we don't need to fix them in this PR but we can use it to estimate future projects/issues). Thanks!

File: src/aspire/operators/polar_ft.py
Comment (garrettwrong) [inline src/aspire/operators/polar_ft.py:118]: I like that you added this check, thanks.  Let's exercise it in a small test to ensure it stays working this way when someone like me that doesn't know the CL code well works on it in the future. Thanks.

File: src/aspire/utils/coor_trans.py
Comment (garrettwrong) [inline src/aspire/utils/coor_trans.py]: fo -> of

File: src/aspire/utils/coor_trans.py
Comment (garrettwrong) [inline src/aspire/utils/coor_trans.py]: Will we still need to have multiple conventions after this PR?

File: src/aspire/utils/coor_trans.py
Comment (garrettwrong) [inline src/aspire/utils/coor_trans.py]: I think you can move the Ri and Ri3 one level up unless you think that impacts the readability too much.

File: src/aspire/utils/coor_trans.py
Comment (garrettwrong) [inline src/aspire/utils/coor_trans.py]: Want to resuse `(np.cross(Ri3, Rj3) / np.linalg.norm(np.cross(Ri3, Rj3))`?

File: src/aspire/utils/coor_trans.py
Comment (garrettwrong) [inline src/aspire/utils/coor_trans.py]: I'm a little confused why we have %360 twice. L332-333 already mods. Then if >180 subtracting 180 in L336-337 should take any value from [180,360) to [0, 180) right?

File: src/aspire/utils/rotation.py
Comment (garrettwrong) [inline src/aspire/utils/rotation.py:414]: Thinking maybe `angle_dist` should handle the stack(s), but I understand doing it this way avoids dealing with broadcasting/stack ambiguity.  Can defer to Joakim on this. I'm fine with what you've done here for the time being.  This (polar 2d ft) already took a little more time than we thought to verify.

File: tests/test_orient_symmetric.py
Comment (garrettwrong) [inline tests/test_orient_symmetric.py]: maybe I missed this before, is this just min(dist)?

File: tests/test_orient_symmetric.py
Comment (garrettwrong) [inline tests/test_orient_symmetric.py:196]: Shame we can't use that new function in either of these two spots (L177, L211).  Perhaps if the angle dist handled stacks might cover one of them.  I'm still thinking kick the can on that, but maybe you see a way to make it work without a lot of effort.

File: tests/test_orient_symmetric.py
Comment (garrettwrong) [inline tests/test_orient_symmetric.py]: I'm sure you can think of something better for `with_<...>`, so that this code is easier to maintian.  For example. looks like we already missed one mating change on L392 (5->2 right?).

BTW, if we improved from 5* to 2*, that is great! Nice job!

File: tests/test_orient_sync_voting.py
Comment (garrettwrong) [inline tests/test_orient_sync_voting.py]: Ballpark, how long does this add to the total unit test time?

File: tests/test_orient_sync_voting.py
Comment (garrettwrong) [inline tests/test_orient_sync_voting.py:106]: Where does the 5 come into play, or was that part of the step_size usage issue(bug)?

File: tests/test_orient_sync_voting.py
Comment (garrettwrong) [inline tests/test_orient_sync_voting.py:6]: Thank you for converting this while you were working this issue!

File: tests/test_polar_ft.py
Comment (garrettwrong) [inline tests/test_polar_ft.py]: compenent->component (and below)

File: tests/test_polar_ft.py
Comment (garrettwrong) [inline tests/test_polar_ft.py]: During review I looked at the following modes from one of the papers I sent you.

```
+    (1, 1),
+    (2, 1),  # (1,2) in Wang, Ronneberger, Burkhardt
+    (5, 1),  # (5,1) ""
+    (1, 2),  # (2,1) ""
+    (3, 2),  # (2,3) ""
```

I think we should try to us them and cite the reference (perhaps instead of the ones you added). The paper seems like a pretty complete publication on the topic, and the generated images can be compared directly to the paper if you add a zero boundary condition to the function generation (try this hack) and plot. Success :)

`* (g["r"] < 1).astype(dtype)  # hack`

Unfortunately, `test_angular_radial_modes` seems to fail for these (regardless of the zero boundary). Hrmmm.   So, it looks like equation 61 in that paper provides a formula to convert from discrete Cartesian FFT to Polar FT.  I haven't tried it, but based on what I _have_ tried from that paper so far I have no reason to believe it wouldn't work.  Maybe that is an easier thing to test?

**Review by garrettwrong — CHANGES_REQUESTED** [review]
Hey, this is great. Thank you for taking the time to clean this up and verify the implementation.

Mostly quick small stuff.  There is the one mode testing issue, but I took a look at it and hope the suggestion is helpful.

File: src/aspire/abinitio/commonline_base.py
Comment (j-c-c) [reply src/aspire/abinitio/commonline_base.py]: Good point. I'll just use the proper reshape. 

Thanks. Done.

File: src/aspire/abinitio/commonline_base.py
Comment (j-c-c) [reply src/aspire/abinitio/commonline_base.py:267]: It was to fix a bug. Estimating shifts requires that rotations have already been estimated. In any case except using a simulation we would get some error do to rotations being `None`. I did come across this error, but can't recall the exact circumstance.

Added a comment for clarification.

File: src/aspire/operators/polar_ft.py
Comment (j-c-c) [reply src/aspire/operators/polar_ft.py]: Sounds good to me. Any suggestion for a name? `full_polar_ft`?

File: src/aspire/operators/polar_ft.py
Comment (j-c-c) [reply src/aspire/operators/polar_ft.py]: Nothing complains. This check and the one on line 108 were previously being handled by `Basis` when this was a subclass.

I'll change the mention of `evaluate_t` since this is now called `transform`.

File: src/aspire/operators/polar_ft.py
Comment (j-c-c) [reply src/aspire/operators/polar_ft.py:103]: Nothing complaining for this one.

File: src/aspire/operators/polar_ft.py
Comment (j-c-c) [reply src/aspire/operators/polar_ft.py:118]: Oops, good call. Added. Thanks

File: src/aspire/utils/coor_trans.py
Comment (j-c-c) [reply src/aspire/utils/coor_trans.py]: Thanks, fixed.

File: src/aspire/utils/coor_trans.py
Comment (j-c-c) [reply src/aspire/utils/coor_trans.py]: No, I'll remove this flag. Thanks

File: src/aspire/utils/coor_trans.py
Comment (j-c-c) [reply src/aspire/utils/coor_trans.py]: Good call. Fixed. Thanks

File: src/aspire/utils/coor_trans.py
Comment (j-c-c) [reply src/aspire/utils/coor_trans.py]: Sure! Done

File: src/aspire/utils/coor_trans.py
Comment (j-c-c) [reply src/aspire/utils/coor_trans.py]: Yep, you're correct. Removed the extra mod for the `ij` index. Still need for `ji` since it could be in [0, 360) prior to subtracting. Thanks

File: tests/test_orient_symmetric.py
Comment (j-c-c) [reply tests/test_orient_symmetric.py]: lol. yep. Not sure what I was thinking there! Fixed.

File: tests/test_orient_symmetric.py
Comment (j-c-c) [reply tests/test_orient_symmetric.py:196]: Yeah, we'd still need the loop to find which rotation the estimate  matches best...

File: tests/test_orient_sync_voting.py
Comment (j-c-c) [reply tests/test_orient_sync_voting.py:6]: 😇

File: tests/test_orient_symmetric.py
Comment (j-c-c) [reply tests/test_orient_symmetric.py:196]: Can use the new function above in `test_estimate_rotations`. Replaced that section with one line.

File: tests/test_orient_symmetric.py
Comment (j-c-c) [reply tests/test_orient_symmetric.py]: Ok, changed to `within_tol`, with an `angle_tol` variable that can be changed.

File: tests/test_orient_sync_voting.py
Comment (j-c-c) [reply tests/test_orient_sync_voting.py:106]: This part of the step size bug. Forgot to remove this. Gone now. Thanks

File: tests/test_orient_sync_voting.py
Comment (j-c-c) [reply tests/test_orient_sync_voting.py:6]: 😇

File: tests/test_orient_sync_voting.py
Comment (j-c-c) [reply tests/test_orient_sync_voting.py]: ~ 1min 45sec under singles. Set doubles to "expensive". 

This is the majority of the unit test time, full test takes just under 2min. I could set the random offsets to "expensive" instead of doubles if you think this is too much added time.

File: tests/test_polar_ft.py
Comment (j-c-c) [reply tests/test_polar_ft.py]: Oops. Fixed. Thanks!

File: tests/test_polar_ft.py
Comment (j-c-c) [reply tests/test_polar_ft.py]: Removed this test for now and will open a follow-up PR.

**Comment by j-c-c** [comment]
Oops, sorry @garrettwrong. Left a `breakpoint()` in there. Should be all good now.

File: src/aspire/operators/polar_ft.py
Comment (garrettwrong) [reply src/aspire/operators/polar_ft.py]: The class is already PolarFT, so `full()` seems okay.

File: src/aspire/operators/polar_ft.py
Comment (garrettwrong) [reply src/aspire/operators/polar_ft.py:103]: If nothing complains why not raise then?

File: src/aspire/operators/polar_ft.py
Comment (garrettwrong) [reply src/aspire/operators/polar_ft.py]: ?

File: src/aspire/operators/polar_ft.py
Comment (garrettwrong) [reply src/aspire/operators/polar_ft.py]: can probably raise then. This code might allow mixed modes, yuk.

File: tests/test_orient_sync_voting.py
Comment (garrettwrong) [reply tests/test_orient_sync_voting.py]: yeah adding almost two minutes is a bit much for no new function.  I think we can keep one case to exercise the code and the others can become expensive. (Should be similar amount of time as before then right?)  Perhaps in the future we can make them faster, but that's another project.

I think it is great to add the coverage (thanks!), just not sure need to exercise it every pytest.

**Review by garrettwrong — CHANGES_REQUESTED** [review]
Still a couple stragglers, but otherwise looking good. Thanks.

File: src/aspire/operators/polar_ft.py
Comment (j-c-c) [reply src/aspire/operators/polar_ft.py]: Oops, don't know how I missed this one! I'll make this change.

File: src/aspire/operators/polar_ft.py
Comment (j-c-c) [reply src/aspire/operators/polar_ft.py]: Ok, added `full()` with a test.

File: src/aspire/operators/polar_ft.py
Comment (j-c-c) [reply src/aspire/operators/polar_ft.py]: Cleaned this up a bit. Thanks.

File: src/aspire/operators/polar_ft.py
Comment (j-c-c) [reply src/aspire/operators/polar_ft.py]: Raising now.

File: src/aspire/operators/polar_ft.py
Comment (j-c-c) [reply src/aspire/operators/polar_ft.py:103]: Changed to raise and added tests.

File: tests/test_orient_sync_voting.py
Comment (j-c-c) [reply tests/test_orient_sync_voting.py]: Ok, marked `None`, ie. random offsets, to be expensive. Also, doubles are marked as expensive. The total unit test time is now ~11sec.

File: src/aspire/operators/polar_ft.py
Comment (garrettwrong) [inline src/aspire/operators/polar_ft.py]: Looks good. I see you didn't like my `pf = full(compact_pf)` naming idea eh 😇 😆 .

File: tests/test_orient_symmetric.py
Comment (garrettwrong) [inline tests/test_orient_symmetric.py:126]: That came out nice, thanks.

File: tests/test_polar_ft.py
Comment (garrettwrong) [inline tests/test_polar_ft.py]: Think we missed removing this line. `pf` is supposed to be this shape already right?

File: src/aspire/operators/polar_ft.py
Comment (garrettwrong) [reply src/aspire/operators/polar_ft.py]: Nice, I can actually read it quickly now. Nice job.

File: tests/test_orient_sync_voting.py
Comment (garrettwrong) [reply tests/test_orient_sync_voting.py]: Nice! Thanks.

**Review by garrettwrong — DISMISSED** [review]
🚀

File: tests/test_polar_ft.py
Comment (j-c-c) [reply tests/test_polar_ft.py]: Ah, good catch. Thanks!

Comment (j-c-c) [reply tests/test_polar_ft.py]: Fixed it up. Thanks

File: src/aspire/operators/polar_ft.py
Comment (j-c-c) [reply src/aspire/operators/polar_ft.py]: I figured the "full" said enough 😇

File: tests/test_orient_symmetric.py
Comment (j-c-c) [reply tests/test_orient_symmetric.py:126]: I thought so too. All cases improved, but I figured 3 degrees was a reasonable tolerance and covered the weakest links.

**Review by garrettwrong — DISMISSED** [review]

**Comment by j-c-c** [comment]
@janden, this is ready for you to take a look whenever you get a chance. Tagging you here since I mistakenly opened this as a PR instead of a draft and you might have not gotten a review request when I marked as "Ready for Review".

File: src/aspire/operators/polar_ft.py
Comment (janden) [inline src/aspire/operators/polar_ft.py]: `2D polar Fourier grid` → `polar Fourier transform`

File: src/aspire/operators/polar_ft.py
Comment (janden) [inline src/aspire/operators/polar_ft.py]: We haven't defined what the “grid” is here (also below, under `dtype`). It should either be clarified or replaced with some other word (here “transform” would work).

File: src/aspire/operators/polar_ft.py
Comment (janden) [inline src/aspire/operators/polar_ft.py]: This doesn't seem to be used anywhere.

File: src/aspire/operators/polar_ft.py
Comment (janden) [inline src/aspire/operators/polar_ft.py]: `.transform`, no?

File: src/aspire/utils/rotation.py
Comment (janden) [inline src/aspire/utils/rotation.py]: Shouldn't this be taking `Rotation` objects as input?

File: src/aspire/utils/rotation.py
Comment (janden) [reply src/aspire/utils/rotation.py:414]: Would vote for having `angle_dist` take care of stacks. (Then this function is just `np.mean(angular_distance(rots1, rots2))`.) This works for now, but let's make an issue.

File: src/aspire/operators/polar_ft.py
Comment (janden) [inline src/aspire/operators/polar_ft.py]: Would it be easier (simpler?) if this was just a kwarg to `transform`. I.e. calling `transform(full=True)` would give the full transform?

**Review by janden — CHANGES_REQUESTED** [review]
Looks good! Just a few things.

File: src/aspire/operators/polar_ft.py
Comment (j-c-c) [reply src/aspire/operators/polar_ft.py]: Thanks. Done.

File: src/aspire/operators/polar_ft.py
Comment (j-c-c) [reply src/aspire/operators/polar_ft.py]: Sounds good. Done.

File: src/aspire/operators/polar_ft.py
Comment (j-c-c) [reply src/aspire/operators/polar_ft.py]: Good catch. Thank you. I was initially implementing a lazy evaluation of the half transform to work with the `full` method. Removed this.

File: src/aspire/operators/polar_ft.py
Comment (j-c-c) [reply src/aspire/operators/polar_ft.py]: Sure, that looks better. These were just copied over.

Done.

File: src/aspire/operators/polar_ft.py
Comment (j-c-c) [reply src/aspire/operators/polar_ft.py]: I'm kind of split on this. We currently have algorithms that get the cross correlation between the half_transform and the full_transform. With this current implementation we would only compute the half_transform and then infer the full_transform using conjugate symmetry with `full(pf)` when needed. This just puts the computation for those uses in one (tested) location. 

If we make this a kwarg of transform we would be recomputing the full transform when calling `transform(full=True)`, which is something we should avoid if possible.  We could cache the half_transform and use it if it exists to compute the full_transform, but I think this begins to make things more complicated than necessary. 

What do you think?

File: src/aspire/operators/polar_ft.py
Comment (janden) [reply src/aspire/operators/polar_ft.py]: Got it.

File: src/aspire/operators/polar_ft.py
Comment (janden) [reply src/aspire/operators/polar_ft.py]: Well if the user wants both full and half, can't they get the full transform and split it?

File: src/aspire/utils/rotation.py
Comment (j-c-c) [reply src/aspire/utils/rotation.py]: This is mainly a diagnostic tool that is used on `src.rotations` which is implemented in `ImageSource` as `self._rotations.matrices`, ie. as a `numpy.ndarray`.

Will add support for `Rotation` objects as well.

File: src/aspire/operators/polar_ft.py
Comment (j-c-c) [reply src/aspire/operators/polar_ft.py]: Sure, they could. The main user is currently our own algorithms. As the algorithms stand now, we mostly use the half and build the full when needed. 

I think ultimately we should just decide if we want `transform` to return the "full" transform or the "half". If it's full then we would store the larger data set and slice it when we need the half. If it's half we store less data and build the full when needed. 

Either way, for the sake of consistency we would probably still want a reusable method to get us from full-to-half or half-to-full (this would be true with or without a kwarg).

File: src/aspire/utils/rotation.py
Comment (j-c-c) [reply src/aspire/utils/rotation.py:414]: Went ahead and just implemented this now. Let me know how it looks.

File: src/aspire/utils/rotation.py
Comment (j-c-c) [reply src/aspire/utils/rotation.py]: Support for `Rotation` objects added as part of having `angle_dist` take care of stacks.

Comment (janden) [reply src/aspire/utils/rotation.py]: Ok cool.

File: src/aspire/utils/rotation.py
Comment (janden) [reply src/aspire/utils/rotation.py:414]: Looks great. Thanks!

File: src/aspire/operators/polar_ft.py
Comment (janden) [reply src/aspire/operators/polar_ft.py]: Ok makes sense, but maybe rename it `half_to_full` so that it's clear that it takes an input and transforms it (since it's a method, it might otherwise look like it does `transform` but returns the full version).

Comment (j-c-c) [reply src/aspire/operators/polar_ft.py]: Ah good point. Will do.

Comment (j-c-c) [reply src/aspire/operators/polar_ft.py]: Done!

File: src/aspire/utils/rotation.py
Comment (j-c-c) [inline src/aspire/utils/rotation.py:370]: Hey @garrettwrong, just tagging you here in case you want to take a look at the changes to this method that were made after you approval.

File: src/aspire/operators/polar_ft.py
Comment (janden) [reply src/aspire/operators/polar_ft.py]: :+1:

**Review by janden — APPROVED** [review]

**Pull request merged**
